### ISSUE TITLE
Adding tests to descriptors

### ISSUE THREAD

**Issue by Trajan64** [boutiques/boutiques#204]
**Quick description**: It will soon be possible to define test-cases within a boutiques descriptor. Each of those test-cases will have associated input information and assertion scenario(s). With the execution of a bosh command (see the usage sub-section), the program will automatically execute and report (to the user) the test-cases found inside the provided descriptor.

# 1) Update to the schema descriptor

The schema descriptor will be modified to include the possibility of defining test-cases.

The descriptor should now include a new “test” property. This property will hold (a possible) array of “test-cases” where each of those test-cases will be defined by:

- A name, 
- An invocation object, which will indicate how to run the test-case (inputs and such).
- An object holding assertions strategies. Those strategies are: (at least one strategy should be included)

   - Matching the returned code of the test-case against an expected exit-code.

   - Checking each of the outputted files against a specified list of path names 
(optionally) and check the content of those said files against reference files.  

Following is a draft of what has been mentioned above. It is written in the form of a JSON Schema chunk which is meant to be introduced to descriptor.schema.json:
(This is WIP; I still have some things to learn about Schema)
	
        
```
   ...
   "tests": {
      "id": "http://github.com/boutiques/boutiques-schema/tests",
      "type": "array",
      "items": {
         "id": "http://github.com/boutiques/boutiques-schema/tests/test-case",
         "type": "object",
         "properties": {
            "name": {
               "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/name",
               "minLength": 1,
               "description": "Name of the test-case",
               "type": "string"
            },
            "invocation":  {
               "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/invocation",
               "type": "object"
            },
            "assertions": {
               "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/assertions",
               "type": "object",
               "properties": {
                  "exit-code": {
                     "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/assertions/exit-code",
                     "description": "Expected code returned by the program.",
                     "type": "integer"
                  }
                  "out-files": {
                     "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/assertions/out-files",
                     "type": "array",
                     "minItems": 1,
                     "items": {
                        "id": "http://github.com/boutiques/boutiques-schema/test-case/assertions/out-files/output",
                        "type": "object",
                        "properties": {
                           "path": {
                              "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/assertions/out-files/output/path",
                              "description": "Expected path of one of the outputted file",
                              "minLength": 1,
                              "type": "string"
                           },
                           "reference": {
                              "id": "http://github.com/boutiques/boutiques-schema/tests/test-case/assertions/out-files/output/reference",
                              "description": "Path to a file for which its content should be matched against a specific outputted file. The outputted file to match should be the one for which its path equals to the one specified in the path field.",
                              "minLength": 1,
                              "type": "string"
                           }
                        },
                        "required": [ "path"]
                     }
                  }
               },
               "oneOf": [
                  { "required": ["exit-code"] },
                  { "required": ["out-files"] }
               ]
            }
         },
      "required": [
         "name",
         "assertions"
      ]
      }
   },
   ...
```

# 2) Structure of a test

## 1. Input data

Input data will be specified as an invocation object, no modifications to the invocation schema is necessary.
	
However, an inconvenient of using the invocation schema as-is is that it forces the user to maintain the testing data locally (as invocation objects deal with file paths and not URL).
To prevent this, the path within the invocation object could point toward virtualized data. 
Using an object container (such as Docker) would effectively enable the usage of URL "disguised" as file paths.
 
## 2. Test evaluation
	
There are 2 different criterias for determining if a test failed or succeeded:

- . Specifying the expected exit code
- . Specifying a list of expected output file(s) (reference(s)). Test failure would happen if either:

   1. (count of expected files) != (count of outputted files)
   2. The location of one of the outputted file is not an element of {locations of the expected files} 
      (cannot find the associated reference)
   3. The content of one of the outputted files does not match the content of its reference.

# 3) Performing tests through bosh

### Command
`bosh test <descriptor.json>     [--max-fail=<INT>] [--print|--print-v] [--order <TEST CASE>+]`

### Options

- `  --print`
Print the results of the test-cases ounto the console. The results will indicate if each test's assertion was a success or a failure.

- `  --print-v`
Produce more information than '-print'. For successful completion of test-cases (no exceptions encountered), the program will output:
(1) Exit-code of the pipeline.
(2) A list of the paths of the outputted files (if any were produced by the pipeline).
Alternatively, if an error was encountered during testing, the program will output details about the associated exception.

- `  --max-fail=<INT>`
Testing will stop after a maximum of x assertion failures.

- `  --order <TEST CASE>+`
Specify the order of execution of the test cases.  One or more (names of) test cases can be specified. 
If this option is not specified, the program will simply assume that all the test-cases must be run in the order by which they appear in the descriptor.
        
### Exit-codes

- Returns 0 if all tests were asserted successfully.
- Returns 1 if one or more tests failed the assertions.
- Returns 2 if no tests were collected in the given descriptor.

## 2. Implementation details
	
The implementation will mainly involve the introduction of a new function to bosh.py: `test(*params)`
The test function will proceed in different steps, it will:

1. Check if the descriptor is valid (globally) 
-> If failure: Raise exception
2. Generate the invocation-schema
3. Extract each of the (defined) invocation objects within the descriptor (inside the test-cases) and validate them against the generated invocation-schema.

	  (If the user specified names of test cases directly through the command then only perform the check-up on those tests and not all of them).
-> If failure: Raise exception.
-> If no invocation objects found: Return with exit-code 2.
4. Sequentially run each of the test-cases and store information about their completions.
	- (Again, if test-cases were specified directly on the command, then only those tests should be run. Additionally they should also be run in the order by which they have been supplied)
	- (If the option --print or --print-v was set: print details on the console of the test that has just completed)
5. Return the appropriate exit-code (see EXIT-CODES above)

**NOTES:**

- From steps (1) to (3), the function will behave much like the `invocation(*params)` function.
The only differences being that: 

  - We may have multiple invocation objects to check against
  - An invocation object will be included inside the descriptor **(Q: Is this okay ? )** and not inside a stand-alone JSON file.

- The execution of a test-case will be done using a LocalExecutor instance, just like how its done in the `execute(*params)` function.

**Comment by glatard** [comment]
Awesome @Trajan64, you seem to be mastering json schema already! 

Minor comments:
* An invocation would also be required in a test object.
* I'm not particularly thrilled by the need for an `order` parameter but that's fine.
* Yes, the invocations will be included in the test objects of the tests array in the descriptor.
* You should also write tests for `bosh test`. 
* `out-files` could be named `output-files`, for consistency.

Looking forward for the first PR! 

@gkiar and @ValHayot might have comments too.

**Comment by gkiar** [comment]
Hey @Trajan64 thanks for the issue; I'm very excited about this, you seem to be a very quick learner! A few comments:

### section 1
looks great - I haven't read the schema closely yet, as I expect this will change as you implement, but the concept seems good.

### section 2.1
I think that in the first iteration, local paths to data (where for containerized apps, "local" means within the image) are sufficient. Subsequently, I don't think it would break the invocation schema to put URLs in the place of filenames and have `bosh test` grab them, store them locally, and convert paths to the `/tmp` locations assigned to the downloaded data.

We may have to increase the intelligence of our validator for the last particular use case, but that is manageable.

### section 3.1
Cool, though I don't love the proposed command-line API as-is, largely because of my thoughts on implementation will I'll mention here and elaborate on below. I think it would be great if the main "thinking" of `bosh test` was to create `pytest` tests for the descriptor and invocations, and then would get `pytest` to run them.

In other words, let `pytest` determine the printing to console, verbosity, how to report number of failures, determine exit codes, and order the tests (this may be the exception, but we can have a `"priority"` property for each test in the descriptor that is used to determine order if needed).

I think this would be a lot less work and be more familiar to users, meaning the command-line would become: `bosh test descriptor.json`.

### section 3.2
Great start on an implementation! You've clearly looked into the code base and tools 😄!

I think an implementation which may require writing less code, and leverage existing standards for testing, might be something like the following pseudocode (note: would use the python api of course, I'm just showing the command-line one where applicable since it's more intuitive).

```
1. bosh validate descriptor.json    # ensure object is real
2. for test in descriptor.tests:    # for every test...
3.    write invocation_testid.json    # write the invocation out to disk
4.    bosh invocation descriptor.json -i invocation_testid.json    # validate the invocation against the invocation schema
5.    create pytest_test_for_this_testid    # dynamically generates a simple command to send pytest for running this test; this will be a lot like those in `test_exec.py` or `test_example1.py`.
5. endfor
6. run pytest on your tests    # do the thing
```

### summary
This is awesome, I'm very excited, and look forward to hearing your thoughts about my proposed implementation and then seeing your PR.

**Comment by glatard** [comment]
just an idea: maybe it would be useful to leverage `pytest` to implement `bosh test`. As in, bosh test would only create pytest tests and use pytest to check them.

**Comment by glatard** [comment]
ok, I just saw that @gkiar already suggested that, sorry!

**Comment by Trajan64** [comment]
@glatard  @gkiar Your feedback is really appreciated, thank to you both! :)

@gkiar Your idea of using the already established functions + pytest to run the tests is very interesting. This would indeed reduce the coding overhead, and possibly bring more functionalities as pytest will offer more commands and options.

If I understand correctly, additional commands and options given to `bosh test` will be used, later on, as arguments to pytest when it will come to properly begin the testing ? 

I suppose I may need to look more in depth into pytest because I don't actually know much about it. I've only briefly checked the usage documentation for the purpose of this report.

Anyway this idea is sure very attractive,. Thanks again to you both :)

**Comment by gkiar** [comment]
Hey @Trajan64 - indeed, additional command-line args could be passed, but I don't think they'd even be necessary as a first pass as the default behaviour of pytest is quite reasonable for most settings, I think. You can think about piping additional args to pytest as a second-level feature, and beyond the scope for this first implementation. We can create an issue with the "enhancement" tag about this :)

**Comment by glatard** [comment]
support for URL could also go in the "enhancement issue".

**Comment by Trajan64** [comment]
@glatard  @gkiar Perfect. :)
I'll proceed as you both said.
Getting something basic but functional.

**Comment by glatard** [comment]
Fixed by @Trajan64 in #214
